dipy/align/mattes.pyx

```diff
@@ -0,0 +1,1390 @@
+        the object of interest and 0's along the background)
+    smin: float
+        the minimum observed intensity associated to the static image, which
+        was used to define the joint PDF
+    sdelta: float
+        bin size associated to the intensities of the static image
+    mmin: float
+        the minimum observed intensity associated to the moving image, which
+        was used to define the joint PDF
+    mdelta: float
+        bin size associated to the intensities of the moving image
+    nbins: int
+        number of histogram bins
+    padding: int
+        number of bins used as padding (the total bins used for padding at both
+        sides of the histogram is actually 2*padding)
+    grad_pdf: array, shape (nbins, nbins, len(theta))
+        the array to write the gradient to
+    '''
+    cdef:
+        int nslices = static.shape[0]
+        int nrows = static.shape[1]
+        int ncols = static.shape[2]
+        int n = theta.shape[0]
+        int offset, valid_points, constant_jacobian=0
+        cnp.npy_intp l, k, i, j, r, c
+        double rn, cn
+        double val, spline_arg
+        double[:,:] J = np.ndarray(shape=(3, n), dtype=np.float64)
+        double[:] prod = np.ndarray(shape=(n,), dtype=np.float64)
+        double[:] x = np.ndarray(shape=(3,), dtype=np.float64)
+
+    grad_pdf[...] = 0
+    with nogil:
+        valid_points = 0
+        for k in range(nslices):
+            for i in range(nrows):
+                for j in range(ncols):
+                    if smask is not None and smask[k, i, j] == 0:
+                        continue
+                    if mmask is not None and mmask[k, i, j] == 0:
+                        continue
+                    valid_points += 1
+                    x[0] = _apply_affine_3d_x0(k, i, j, 1, grid_to_space)
+                    x[1] = _apply_affine_3d_x1(k, i, j, 1, grid_to_space)
+                    x[2] = _apply_affine_3d_x2(k, i, j, 1, grid_to_space)
+
+                    if constant_jacobian == 0:
+                        constant_jacobian = transform._jacobian(theta, x, J)
+
+                    for l in range(n):
+                        prod[l] = J[0,l] * mgradient[k,i,j,0] +\
+                                  J[1,l] * mgradient[k,i,j,1] +\
+                                  J[2,l] * mgradient[k,i,j,2]
+
+                    rn = _bin_normalize(static[k, i, j], smin, sdelta)
+                    r = _bin_index(rn, nbins, padding)
+                    cn = _bin_normalize(moving[k, i, j], mmin, mdelta)
+                    c = _bin_index(cn, nbins, padding)
+                    spline_arg = (c-2) - cn
+
+                    for offset in range(-2, 3):
+                        val = _cubic_spline_derivative(spline_arg)
+                        for l in range(n):
+                            grad_pdf[r, c + offset,l] -= val * prod[l]
+                        spline_arg += 1.0
+
+        if valid_points * mdelta > 0:
```

Comment:
Yes! much better, and more efficient I guess (I wonder if the compiler optimizes that kind of things to avoid doing the multiplication at each iteration). Fixed in all the file.